src/databases/BOV/avtBOVFileFormat.C

```diff
@@ -42,7 +42,7 @@
 
 #include <avtBOVFileFormat.h>
 
-#include <vtk_zlib.h>
+#include <zlib.h>
```

Comment:
Shouldn't these be conditioned like so...

```
#ifdef HAVE_ZLIB_H
#include <zlib.h>
#endif

Reply:
What about the other sections of the plugin code that require zlib. Should they all be conditioned, or should the entire plugin be conditioned on the presence of zlib in the databases/CMakeLists.txt?

Reply:
Depends. Does the plugin not function *at all* without zlib or does the plugin just loose some features? If the former, disable plugin entirely. Otherwise, condition zlib.h inclusion.

Reply:
I think BOV plugin in its entirety should depend on zlib being found. It can open files with '.gz' extension. Unless we add a warning to the plugin when user attempts to open a gzipped file and we don't have zlib support.  Plus there are a couple sections in the writer where I'm not certain the file will be written properly without zlib. See section of writer starting at line 719.

Reply:
When you say it *can* open files with `.gz` extension, do you mean it indeed *requires* them to be `.gz` (gzip compressed) files and there is no coding to allow it to read data from UNcompressed files. If so, I agree that when zlib is NOT enabled, the BOV plugin should also be NOT enabled.

Reply:
I am working through the code that uses gz. Easy enough to wrap it with preprocessor directives if we don't have zlib.  However, I am not familiar enough with databases to know when to bail out and issue a warning/error if BOV should be able to read a gz file but cannot because zlib wasn't found. It looks like only the GetVar method looks for the extension.

Reply:
Ok, so the use case is...a user goes to open a compressed BOV file and VisIt fails to open it because the VisIt they are using was NOT compiled with zlib support. Is that right? And, you'd like to issue a descriptive error message instead of silently failing? 

I think the only way can detect that condition in the plugin is if you read the first few bytes of the file and they match (one of) the zlib header magic number, and this is not a zlib build, you can issue an error message. Maybe more trouble than its worth though.

Reply:
Yes, that was my thought, give the user some indication as to why they cannot open their compressed BOV file.

Reply:
Ok, here's what I would do. Around [this line](https://github.com/visit-dav/visit/blob/37262d6cf642a3b0f0b29b11919989389c00ca6e/src/databases/BOV/avtBOVFileFormat.C#L1188), add logic similar to following

```
 size_t nread = fread(rv->GetVoidPointer(0), unit_size,
                                 nvals * dataNumComponents, file_handle);

if (nread != unit_size * nvals * dataNumComponents) // fread failed in some way
{
    unsigned short magic = 0x0000;
    fread(&magic, 2, 1, file_handle);
    if (magic == 0x1F8B || magic == 0x8B1F) // avert endieness issues by checking both orders
    {
        // Probable attempt to read compressed BOV data with a plugin that was not compiled with zlib enabled.
    }
}
```

There are various places gzip magic number is described. [Here's one](https://en.wikipedia.org/wiki/List_of_file_signatures) -- search for `gzip`.

There are two places in the file to put this check so maybe create a local `avtBOVFileFormat::Read()` method to encapsulate it.

Reply:
oops, you need a `fseek()` in there too to rewind the file back to beginning *before* doing the `fread` for magic number.

Reply:
Isn't the GetVar method a little late in the process to be bailing on the reader?

Reply:
I don't think so. Can you identify an earlier point? Note...BOV header is uncompressed ascii. There is no point in compressing this small file. If BOV data is compressed, only the variable data will be. The reader can become aware of a problem related to compression *only* after its made an attempt to read a compressed file. And, we're a bit lucky in this regard because each variable is in its *own* file so a *short* read is detectable. If multiple variables were in the same file, a *short* read would not be detectable because the read would read past the first compressed variable and into the others and we'd be none the wizer. Come to think of it, this *is* a problem even for a single variable per file because some variable's *compressed* file may in fact wind up being *larger* than the uncompressed data. We cannot detect a *short* read in that case either. That would mean VisIt would read the data, thinking is it fine and all hell would break loose (probably with a lot of NaNs) later with no clue as to why.

Without detecting a read failure, the only option we have left is to inspect the first two bytes of the variable file for tell-tail signs its gzip'd (gzip magic number), which I think is potentially problematic because we could have real float/double binary data at the zero offset in the file that would match that, I think.

So, I think it *has* to be the combination of a read failure *and* that magic before we can reasonably assume its compressed data.

I do not see anything in the BOV header that can be used to explicitly indicate if data is compressed or not (even though our Getting Data Into VisIt Manual in fact mentions that as an option). However, since BOV appears to be a format we/VisIt created, you could decide to add such a keyword there. But, that would effect only new BOV data created going forward and not any existing BOV data.

Another route to avoid all these issues entirely is to just decide to always build VisIt with zlib support and I think that would be a fine idea.

Reply:
So, make zlib a required library in build_visit, avoiding use of system zlib?
Or, still allow build_visit to use either system zlib or one built with build_visit.

For VisIt itself, make zlib a required library such that if FindZlib fails to find either a system or built version of zlib, configure cannot continue?

Or, just make the plugins in question not be built if we don't have zlib?  

I think in most cases (for most platforms?), we will find zlib, either the one built with build_visit or a system version.

Reply:
Not sure. AFAICT, only risk in *requiring* zlib is chance that it won't build somewhere we need it. I think it is highly unlikely that zlib would NOT build where we need it but rest of VisIt would build...which would mean zlib is killing our ability to build somewhere.

Since zlib is so ubiquitous, in 99% of the cases, *requiring* it would mean just adding `-lz` to `LDFLAGS`.

Just another avenue to consider and which might simplify many of these issues.